internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -0,0 +1,499 @@
+	ctx := context.Background()
+
+	// No need to reconcile if the image URL has been set in
+	// the BMH already.
+	if bmh.Spec.Image != nil && bmh.Spec.Image.URL != "" {
+		return reconcileComplete{}
+	}
+
+	r.Log.Debugf("Started BMH reconcile for %s/%s", bmh.Namespace, bmh.Name)
+
+	for ann, value := range bmh.Labels {
+
+		// Find the `BMH_INSTALL_ENV_LABEL`, get the isntallEnv configured in it
+		// and copy the ISO Url from the InstallEnv to the BMH resource.
+		if strings.HasPrefix(ann, BMH_INSTALL_ENV_LABEL) {
+			installEnv := &adiiov1alpha1.InstallEnv{}
+			if err := r.Get(ctx, types.NamespacedName{Name: value, Namespace: bmh.Namespace}, installEnv); err != nil {
+				return reconcileError{client.IgnoreNotFound(err)}
+			}
+
+			if installEnv.Status.ISODownloadURL == "" {
+				// the image has not been created yet, try later.
+				r.Log.Infof("Image URL for InstallEnv (%s/%s) not available yet. Retrying reconcile for BareMetalHost  %s/%s",
+					installEnv.Namespace, installEnv.Name, bmh.Namespace, bmh.Name)
+				return reconcileRequeue{time.Minute}
+			}
+
+			if bmh.Spec.Image == nil {
+				bmh.Spec.Image = &bmh_v1alpha1.Image{}
+			}
+
+			liveIso := "live-iso"
+			bmh.Spec.Image.URL = installEnv.Status.ISODownloadURL
+			bmh.Spec.Image.DiskFormat = &liveIso
+
+			// Let's make sure inspection is disabled for BMH resources
+			// that are associated with an agent-based deployment.
+			//
+			// Ideally, the user would do this while creating the BMH
+			// we are just taking extra care that this is the case.
+			if bmh.ObjectMeta.Annotations == nil {
+				bmh.ObjectMeta.Annotations = make(map[string]string)
+			}
+
+			bmh.ObjectMeta.Annotations[BMH_INSPECT_ANNOTATION] = "disabled"
+
+			r.Log.Infof("Image URL has been set in the BareMetalHost  %s/%s", bmh.Namespace, bmh.Name)
+			return reconcileComplete{true}
+		}
+	}
+
+	return reconcileComplete{}
+}
+
+// Finds the agents related to this BMH
+//
+// The BMH <-> agent relation should be one-to-one. This
+// function returns the agent that matches the following
+// criteria.
+//
+// An agent is only related to the BMH if it has an iface
+// whose MacAddress matches the BootMACAddress set in the BMH.
+//
+// In the scenario where more than one Agent match the BootMACAdress
+// from the BMH, the following logic applies:
+//
+// 1. Return the agent that has been approved, if any.
+// 2. If the above fails, return the first agent found.
+//
+// Otherwise, this function will return nil.
+//
+// Eventually, this will be changed to inspect the agents status
+// instead of `Approved`. We want to reconcile only agents that
+// are in specific states (like discovering). Waiting on MGMT-4266
+func (r *BMACReconciler) findAgent(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost) *adiiov1alpha1.Agent {
+	agentList := adiiov1alpha1.AgentList{}
+	err := r.Client.List(ctx, &agentList, client.InNamespace(bmh.Namespace))
+	if err != nil {
+		return nil
+	}
+
+	agents := []*adiiov1alpha1.Agent{}
+
```

Comment:
maybe better to have agentResult := adiiov1alpha1.Agent{} or nil and have  in line 432
`if agent == nil:
    agentResult = agent
`
then there will be a need only for return agentResult 
?

Reply:
This could also work, yeah! I kept the slice because we may be doing different checks on the agents once the work on exposing the agent status is done. So, we may just rewrite this function then.